File: mobile/lib/core/models/fuel_log.dart
class FuelLog {
  final String id;
  final String bikeId;
  final String loggedAt;
  final double odometerReading;
  final double fuelQuantity;
  final double fuelPricePerUnit;
  final double totalCost;
  final bool isFullTank;
  final String? stationName;
  final String? notes;
  final double? kmSinceLast;
  final double? fuelEfficiency;
  final String createdAt;

  const FuelLog({
    required this.id,
    required this.bikeId,
    required this.loggedAt,
    required this.odometerReading,
    required this.fuelQuantity,
    required this.fuelPricePerUnit,
    required this.totalCost,
    this.isFullTank = true,
    this.stationName,
    this.notes,
    this.kmSinceLast,
    this.fuelEfficiency,
    required this.createdAt,
  });

  factory FuelLog.fromJson(Map<String, dynamic> json) {
    return FuelLog(
      id: json['id']?.toString() ?? '',
      bikeId: json['bike_id']?.toString() ?? '',
      loggedAt: json['logged_at'] as String? ?? '',
      odometerReading: (json['odometer_reading'] as num?)?.toDouble() ?? 0.0,
      fuelQuantity: (json['fuel_quantity'] as num?)?.toDouble() ?? 0.0,
      fuelPricePerUnit:
          (json['fuel_price_per_unit'] as num?)?.toDouble() ?? 0.0,
      totalCost: (json['total_cost'] as num?)?.toDouble() ?? 0.0,
      isFullTank: json['is_full_tank'] as bool? ?? true,
      stationName: json['station_name'] as String?,
      notes: json['notes'] as String?,
      kmSinceLast: (json['km_since_last'] as num?)?.toDouble(),
      fuelEfficiency: (json['fuel_efficiency'] as num?)?.toDouble(),
      createdAt: json['created_at'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'bike_id': bikeId,
      'logged_at': loggedAt,
      'odometer_reading': odometerReading,
      'fuel_quantity': fuelQuantity,
      'fuel_price_per_unit': fuelPricePerUnit,
      'total_cost': totalCost,
      'is_full_tank': isFullTank,
      if (stationName != null) 'station_name': stationName,
      if (notes != null) 'notes': notes,
      if (kmSinceLast != null) 'km_since_last': kmSinceLast,
      if (fuelEfficiency != null) 'fuel_efficiency': fuelEfficiency,
      'created_at': createdAt,
    };
  }

  Map<String, dynamic> toCreateJson() {
    return {
      'bike_id': bikeId,
      'logged_at': loggedAt,
      'odometer_reading': odometerReading,
      'fuel_quantity': fuelQuantity,
      'fuel_price_per_unit': fuelPricePerUnit,
      'total_cost': totalCost,
      'is_full_tank': isFullTank,
      if (stationName != null) 'station_name': stationName,
      if (notes != null) 'notes': notes,
    };
  }
}

File: backend/app/models/fuel_log.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import String, Float, Boolean, DateTime, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.base import UUIDMixin, TimestampMixin


class FuelLog(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "fuel_logs"

    bike_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("bikes.id", ondelete="CASCADE"), nullable=False, index=True
    )
    logged_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, index=True)
    odometer_reading: Mapped[float] = mapped_column(Float, nullable=False)
    fuel_quantity: Mapped[float] = mapped_column(Float, nullable=False)
    fuel_price_per_unit: Mapped[float] = mapped_column(Float, nullable=False)
    total_cost: Mapped[float] = mapped_column(Float, nullable=False)
    is_full_tank: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    station_name: Mapped[Optional[str]] = mapped_column(String(200))
    notes: Mapped[Optional[str]] = mapped_column(String(500))

    bike: Mapped["Bike"] = relationship("Bike", back_populates="fuel_logs")
